hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/supervisor/process/ConsumerProcessSupervisor.java

```diff
@@ -0,0 +1,156 @@
+                start(signal.getTarget(), signal.getPayload());
+                break;
+            case RETRANSMIT:
+                process(signal).accept(signal);
+                break;
+            case UPDATE_SUBSCRIPTION:
+            case UPDATE_TOPIC:
+                process(signal).accept(signal);
+                break;
+            case RESTART:
+            case KILL_RESTART:
+                kill(process(signal));
+                taskQueue.offer(Signal.of(Signal.SignalType.START, signal.getTarget()));
+                break;
+            case STOP:
+                process(signal).accept(signal);
+                taskQueue.offer(Signal.of(Signal.SignalType.CLEANUP, signal.getTarget()));
+                break;
+            case CLEANUP:
+                cleanup(process(signal));
+                break;
+            default:
+                break;
+        }
+    }
+
+    private ConsumerProcess process(Signal signal) {
+        return runningProcesses.getProcess(signal.getTarget());
+    }
+
+    private void kill(ConsumerProcess consumerProcess) {
+        logger.info("Interrupting consumer process {}", consumerProcess);
+        Future task = runningProcesses.getExecutionHandle(consumerProcess);
+        if (!task.isDone()) {
+            if (task.cancel(true)) {
+                logger.info("Interrupted consumer process {}", consumerProcess);
+            } else {
+                logger.error("Failed to interrupt consumer process {}, possible stale consumer", consumerProcess);
+            }
+        } else {
+            logger.info("Consumer was already dead process {}", consumerProcess);
+        }
+    }
+
+    private boolean isHealthy(ConsumerProcess consumerProcess) {
+        long delta = clock.millis() - consumerProcess.healtcheckRefreshTime();
+        if (delta > unhealthyAfter) {
+            logger.info("Lost contact with consumer {}, last seen {}ms ago", consumerProcess, delta);
+            return false;
+        }
+        return true;
+    }
+
+    private void start(SubscriptionName subscriptionName, Optional<Consumer> consumer) {
+        logger.info("Starting consumer process {}", subscriptionName);
+
+        ConsumerProcess process;
+        if (consumer.isPresent()) {
+            process = new ConsumerProcess(subscriptionName, consumer.get(), retransmitter, clock);
+        } else if (runningProcesses.hasProcess(subscriptionName)) {
+            process = runningProcesses.getProcess(subscriptionName);
+        }
+        else {
+            logger.error("Tryng to start process for {} without known Consumer, aborting", subscriptionName);
+            return;
+        }
+
+        Future future = executor.execute(process);
+        runningProcesses.add(process, future);
+
+        logger.info("Started consumer process {}", process);
+    }
+
+    private void cleanup(ConsumerProcess consumerProcess) {
+        kill(consumerProcess);
+        runningProcesses.remove(consumerProcess);
```

Comment:
But from the other hand we have running subscription and some stale process instead of having just stale process...